crates/bevy_animation/src/animator.rs

```diff
@@ -0,0 +1,1411 @@
+use std::{
+    any::{type_name, Any, TypeId},
+    borrow::Cow,
+    collections::{HashMap, HashSet},
+    convert::TryFrom,
+    convert::TryInto,
+    fmt,
+};
+
+use anyhow::Result;
+use bevy_app::prelude::{EventReader, Events};
+use bevy_asset::{AssetEvent, Assets, Handle};
+use bevy_core::prelude::*;
+use bevy_ecs::prelude::*;
+use bevy_math::prelude::*;
+use bevy_reflect::{Reflect, ReflectComponent, TypeUuid};
+use bevy_transform::prelude::*;
+use fnv::FnvBuildHasher;
+use smallvec::{smallvec, SmallVec};
+use tracing::warn;
+
+use crate::{
+    blending::AnimatorBlending,
+    hierarchy::Hierarchy,
+    interpolate::Lerp,
+    tracks::ArrayN,
+    tracks::{Track, TrackBase, TrackFixed, TrackFixedN, TrackNBase, ValueN},
+    wide::{Quatx8, Vec3x8},
+};
+
+// TODO: Load Clip name from gltf
+
+// Starting from Intel's Sandy Bridge, spatial prefetcher is now pulling pairs of 64-byte cache
+// lines at a time, so we have to align to 128 bytes rather than 64.
+//
+// Sources:
+// - https://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-optimization-manual.pdf
+// - https://github.com/facebook/folly/blob/1b5288e6eea6df074758f877c849b6e73bbb9fbb/folly/lang/Align.h#L107
+//
+// ARM's big.LITTLE architecture has asymmetric cores and "big" cores have 128 byte cache line size
+// Sources:
+// - https://www.mono-project.com/news/2016/09/12/arm64-icache/
+//
+#[cfg(any(target_arch = "x86_64", target_arch = "aarch64"))]
+const CACHE_SIZE: usize = 128;
+#[cfg(not(any(target_arch = "x86_64", target_arch = "aarch64")))]
+const CACHE_SIZE: usize = 64;
+
+/// Defines the number of keyframes that fits inside a cache line;
+const KEYFRAMES_PER_CACHE: usize = CACHE_SIZE / std::mem::size_of::<u16>();
+
+///////////////////////////////////////////////////////////////////////////////
+
+pub struct Tracks<T> {
```

Comment:
I think the hierarchy of tracks, clips, key frames, layers is a bit lost on me/hard to follow.

Would you mind laying that structure out a bit more clearly in the PR? Also from the PR it seems like a user would mostly just interact with clips. (Thought I feel like key frames and layers may just be as important)

Reply:
Clip is made of a collection of Tracks, each Track animates a single property in a single entity; The animator works by layering multiple Clips on top of each other in order to evaluate the final animation; So each Layer have information about how to play  a single Clip, like time and weight as well some other complicated stuff to make everything work;

Made more sence?

Reply:
Yes that helped a lot =) Thank you very much.